lib/ServiceFileProcessor.js

```diff
@@ -0,0 +1,136 @@
+
+class ServiceFileProcessor {
+  constructor(options) {
+    this.cliManager = options.cliManager;
+    this.servicesService = options.servicesService;
+  }
+
+  process(options, done) {
+    const operationType = options.operation;
+    const source = options.parsedData;
+    if (operationType === OperationType.CREATE) {
+      this._createService(source, options, done);
+    } else if (operationType === OperationType.UPDATE) {
+      this._updateService(source, options, done);
+    } else {
+      return setImmediate(() => { done(new Error(`Operation type not supported: ${operationType}`)); });
+    }
+  }
+
+  _createService(source, options, done) {
+    const data = {
+      name: options.name,
+      type: source.type === 'flex-internal' ? 'internal' : 'external',
+      backingServers: [{ secret: source.secret }]
+    };
+
+    if (!isNullOrUndefined(source.description)) {
+      data.description = source.description;
+    }
+
+    if (source.type === 'flex-external') {
+      data.backingServers[0].host = source.host;
+    }
+
+    this.servicesService.create(data, options.domainId, options.domainType, (err, service) => {
+      if (err) {
+        return done(err);
+      }
+
+      const dir = source.sourcePath;
+      if (!dir) {
+        return done(null, service);
+      }
+
+      this.servicesService.deployFlexProject(dir, service.id, this.cliManager.config.defaultSchemaVersion, (err, id) => {
+        if (err) {
+          return done(err);
+        }
+
+
+        service.jobId = id;
+        done(null, service);
+      });
+    });
+  }
+
+  _updateService(source, options, done) {
+    const serviceId = options.serviceId;
+    let updateData;
+    let jobId;
+
+    async.series([
+      (next) => {
+        this.servicesService.getById(serviceId, (err, originalService) => {
+          if (err) {
+            return next(err);
+          }
+
+          const data = {
+            name: source.name || originalService.name,
+            type: originalService.type, // cannot be updated but required from the backend
+            backingServers: [
+              {
+                host: source.host || originalService.backingServers[0].host, // can be updated for external
+                secret: source.secret || originalService.backingServers[0].secret,
+                _id: originalService.backingServers[0]._id // if not passed, a new id is assigned
+              }
+            ]
+          };
+          if (!isNullOrUndefined(source.description)) {
+            data.description = source.description;
+          }
+
+          updateData = data;
+
+          next();
+        });
+      },
+      (next) => {
+        this.servicesService.update(updateData, serviceId, next);
+      },
+      (next) => {
+        const dir = source.sourcePath;
+        if (!dir) {
+          return setImmediate(next);
+        }
+
+        this.servicesService.deployFlexProject(dir, serviceId, this.cliManager.config.defaultSchemaVersion, (err, id) => {
```

Comment:
Yes, it is possible but to me the question is 'Should we?'. I have a few concerns here:
* we don't have this kind of enhancements/behavior for the other entities - for example for a `push` command if you have specified a certain already existent collection, then we try to update it. In general, if an entity is specified, we behave more like a PUT request than a PATCH (same as the backend).
* with the above in mind, wouldn't a discrepancy emerge if we started skipping only deploy even if `sourcePath` is set in config file?
* it should also be noted that the backend checks if the project's version is already deployed. So if you are doing a deploy in the regular way (`kinvey flex deploy`) and you try to deploy an already deployed version, then this will result in error.

...I guess here we can log a warning that this version is already deployed and spare the user the backend's error, and continue applying the file. Still, I feel that we should have a general strategy/resolution for these kind of things.

Reply:
I agree that ideally we should have similar checks for everything we update. However, I would like us to not do the request for deploy at all, precisely because of the error that we will inevitably get (since the version is not changed). Also, deploy is a much more complex operation (both client- and server-wise), it would slow down the whole operation unnecessarily.